misc/python/materialize/parallel_workload/database.py

```diff
@@ -0,0 +1,150 @@
+
+MAX_COLUMNS = 20
+MAX_INITIAL_TABLES = 10
+MAX_INITIAL_VIEWS = 10
+
+
+class Column:
+    column_id: int
+    data_type: Type[DataType]
+    table: "Table"
+    nullable: bool
+    default: Optional[str]
+
+    def __init__(
+        self,
+        rng: random.Random,
+        column_id: int,
+        data_type: Type[DataType],
+        table: "Table",
+    ):
+        self.column_id = column_id
+        self.data_type = data_type
+        self.table = table
+        self.nullable = rng.choice([True, False])
+        self.default = rng.choice([None, str(data_type.value(rng))])
+
+    def __str__(self) -> str:
+        return f"c{self.column_id}_{self.data_type.name()}"
+
+    def value(self, rng: random.Random) -> str:
+        return str(self.data_type.value(rng))
+
+    def create(self) -> str:
+        result = f"{self} {self.data_type.name()}"
+        if self.default:
+            result += f" DEFAULT {self.default}"
+        if not self.nullable:
+            result += " NOT NULL"
+        return result
+
+
+class DBObject:
+    columns: List[Column]
+
+
+class Table(DBObject):
+    table_id: int
+    rename: int
+
+    def __init__(self, rng: random.Random, table_id: int):
+        self.table_id = table_id
+        self.columns = [
+            Column(rng, i, rng.choice(DATA_TYPES), self)
+            for i in range(rng.randint(2, MAX_COLUMNS))
+        ]
+        self.rename = 0
+
+    def __str__(self) -> str:
+        if self.rename:
+            return f"t{self.table_id}_{self.rename}"
+        return f"t{self.table_id}"
+
+    def create(self, cur: pg8000.Cursor) -> bool:
+        query = f"CREATE TABLE {self}("
+        query += ",\n    ".join(column.create() for column in self.columns)
+        query += ")"
+        return execute(cur, query)
+
+
+class View(DBObject):
+    view_id: int
+    table: DBObject
+    materialized: bool
+
+    def __init__(self, rng: random.Random, view_id: int, table: DBObject):
+        self.view_id = view_id
+        self.table = table
+        self.columns = [
+            column
+            for column in rng.sample(
+                table.columns, k=rng.randint(1, len(table.columns))
+            )
+        ]
+        self.materialized = rng.choice([True, False])
+
+    def __str__(self) -> str:
+        return f"v{self.view_id}"
+
+    def create(self, cur: pg8000.Cursor) -> bool:
+        if self.materialized:
+            query = "CREATE MATERIALIZED VIEW"
+        else:
+            query = "CREATE VIEW"
+        columns = ", ".join(str(column) for column in self.columns)
+        query += f" {self} AS SELECT {columns} FROM {self.table}"
+        return execute(cur, query)
+
+
+class Database:
+    seed: str
+    tables: List[Table]
+    views: List[View]
+    indexes: Set[str]
+    lock: threading.Lock
+
+    def __init__(self, rng: random.Random, seed: str):
+        self.seed = seed
+
+        self.tables = [Table(rng, i) for i in range(rng.randint(2, MAX_INITIAL_TABLES))]
```

Comment:
If possible, in a  DML workload (or maybe in all workloads, by default) all the initial database objects should have an index applied to them by default at start. Otherwise SELECTs are very slow and the workload is not stressful.